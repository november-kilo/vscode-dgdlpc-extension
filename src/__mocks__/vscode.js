class DiagnosticClass {
	constructor(range, message, severity) {
		this.range = range;
		this.message = message;
		this.severity = severity;
		this.code = undefined;
		this.data = undefined;
	}
}

const DiagnosticMock = jest.fn().mockImplementation((range, message, severity) => {
	return new DiagnosticClass(range, message, severity);
});
// Make the prototype chain work correctly for instanceof checks
Object.setPrototypeOf(DiagnosticClass.prototype, DiagnosticMock.prototype);

class LocationClass {
	constructor(uri, position) {
		this.uri = uri;
		this.range = {
			start: position,
			end: position
		};
	}
}

const LocationMock = jest.fn().mockImplementation((uri, position) => {
	return new LocationClass(uri, position);
});

Object.setPrototypeOf(LocationClass.prototype, LocationMock.prototype);

const HoverMock = jest.fn().mockImplementation((contents, range) => ({
	contents: contents ? (Array.isArray(contents) ? contents : [contents]) : [],
	range: range || null
}));

class WorkspaceEditMock {
	constructor() {
		this.replacements = [];
	}

	replace(uri, range, newText) {
		this.replacements.push({ uri, range, newText });
	}
}

class SemanticTokensLegendClass {
	constructor(tokenTypes, tokenModifiers) {
		this.tokenTypes = tokenTypes;
		this.tokenModifiers = tokenModifiers;
	}
}

const SemanticTokensLegendMock = jest.fn().mockImplementation((tokenTypes, tokenModifiers) => {
	return new SemanticTokensLegendClass(tokenTypes, tokenModifiers);
});

// Make the prototype chain work correctly for instanceof checks
Object.setPrototypeOf(SemanticTokensLegendClass.prototype, SemanticTokensLegendMock.prototype);

module.exports = {
	Range: jest.fn((startLine, startChar, endLine, endChar) => ({
		start: { line: startLine, character: startChar },
		end: { line: endLine, character: endChar }
	})),
	Position: jest.fn((line, character) => ({
		line,
		character,
		isAfter: function(other) {
			return this.line > other.line ||
				(this.line === other.line && this.character > other.character);
		},
		isBefore: function(other) {
			return this.line < other.line ||
				(this.line === other.line && this.character < other.character);
		},
		isEqual: function(other) {
			return this.line === other.line && this.character === other.character;
		},
		translate: function(lineDelta, characterDelta) {
			return new Position(
				this.line + (lineDelta || 0),
				this.character + (characterDelta || 0)
			);
		},
		with: function(line, character) {
			return new Position(
				line === undefined ? this.line : line,
				character === undefined ? this.character : character
			);
		}
	})),
	Diagnostic: DiagnosticMock,
	DiagnosticSeverity: {
		Error: 0,
		Warning: 1,
		Information: 2,
		Hint: 3
	},
	CompletionItem: jest.fn().mockImplementation((label, kind) => ({
		label,
		kind,
		documentation: undefined,
		detail: undefined,
		insertText: undefined
	})),
	CompletionItemKind: {
		Function: 'Function',
		Class: 'Class',
		Keyword: 'Keyword'
	},
	MarkdownString: jest.fn().mockImplementation(function(value = '') {
		this.value = value;
		this.appendMarkdown = jest.fn();
		this.isTrusted = false;
		this.supportThemeIcons = false;
		this.supportHtml = false;
		this.appendMarkdown = jest.fn(text => {
			this.value += text;
			return this;
		});
		this.appendCodeblock = jest.fn((code, language) => {
			this.value += `\`\`\`${language}\n${code}\n\`\`\``;
			return this;
		});
	}),
	SnippetString: jest.fn().mockImplementation(function(value = '') {
		this.value = value;
	}),
	CodeAction: jest.fn().mockImplementation((title, kind) => ({
		title,
		kind,
		edit: undefined,
		isPreferred: false,
		diagnostics: []
	})),
	CodeActionKind: {
		QuickFix: 'quickfix'
	},
	window: {
		createOutputChannel: jest.fn(() => ({
			appendLine: jest.fn(),
			dispose: jest.fn()
		})),
		showInputBox: jest.fn(),
		showErrorMessage: jest.fn()
	},
	workspace: {
		getConfiguration: jest.fn().mockImplementation(() => ({
			get: jest.fn()
		})),
		workspaceFolders: [{
			uri: {
				fsPath: '/test/workspace'
			}
		}],
		onDidChangeTextDocument: jest.fn().mockImplementation((callback) => ({
			dispose: jest.fn()
		})),
		onDidCloseTextDocument: jest.fn().mockImplementation((callback) => ({
			dispose: jest.fn()
		})),
		textDocuments: []
	},
	WorkspaceEdit: jest.fn().mockImplementation(() => new WorkspaceEditMock()),
	Uri: {
		file: jest.fn(path => ({ path, scheme: 'file' })),
		parse: jest.fn(uri => ({ uri, scheme: 'file' }))
	},
	Hover: HoverMock,
	commands: {
		registerCommand: jest.fn().mockImplementation((command, callback) => ({
			dispose: jest.fn()
		}))
	},
	SemanticTokensLegend: SemanticTokensLegendMock,
	SemanticTokensBuilder: jest.fn().mockImplementation(() => ({
		push: jest.fn(),
		build: jest.fn().mockReturnValue('built-tokens')
	})),
	languages: {
		createDiagnosticCollection: jest.fn().mockImplementation((name) => ({
			name,
			set: jest.fn(),
			delete: jest.fn(),
			clear: jest.fn(),
			dispose: jest.fn()
		})),
		registerCodeActionsProvider: jest.fn().mockImplementation((selector, provider, metadata) => ({
			dispose: jest.fn()
		})),
		registerCompletionItemProvider: jest.fn().mockImplementation((selector, provider, ...triggerCharacters) => ({
			dispose: jest.fn()
		})),
		registerHoverProvider: jest.fn().mockImplementation((selector, provider) => ({
			dispose: jest.fn()
		})),
		registerDocumentSemanticTokensProvider: jest.fn().mockReturnValue({ dispose: jest.fn() })
	},
	Location: LocationMock
};
