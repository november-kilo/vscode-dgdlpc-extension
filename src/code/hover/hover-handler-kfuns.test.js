import * as vscode from 'vscode';
import KfunsHoverHandler from "./hover-handler-kfuns";
import MarkdownUtil from '../markdown-util';

jest.mock('../markdown-util', () => ({
	bold: jest.fn((text) => `**${text}:**`),
	showDocument: jest.fn((funcName) =>
		`[${funcName}](command:lpc.showKfunDoc?${encodeURIComponent(JSON.stringify([funcName]))})`)
}));

jest.mock('../../kfuns.json', () => ({
	kfuns: {
		'test_fun': {
			description: 'Test function description',
			params: [
				{
					label: 'param1',
					documentation: 'First parameter',
					optional: false
				},
				{
					label: 'param2',
					documentation: 'Second parameter',
					optional: true
				}
			],
			seeAlso: ['related_fun', 'another_fun']
		},
		'no_params_fun': {
			description: 'Function without parameters'
		}
	}
}));

describe('KfunsHoverHandler', () => {
	let handler;
	let mockDocument;
	let mockPosition;
	let mockRange;

	beforeEach(() => {
		handler = new KfunsHoverHandler();
		mockRange = new vscode.Range(0, 0, 0, 8);
		mockPosition = new vscode.Position(0, 5);
		mockDocument = {
			getWordRangeAtPosition: jest.fn().mockReturnValue(mockRange),
			getText: jest.fn()
		};
		jest.clearAllMocks();
	});

	describe('canHandle', () => {
		test('returns false for no range', () => {
			mockDocument.getWordRangeAtPosition.mockReturnValue(null);

			const canHandle = handler.canHandle(mockDocument, mockPosition);

			expect(canHandle).toBe(false);
		});

		test('returns false for unknown kfun', () => {
			mockDocument.getWordRangeAtPosition.mockReturnValue(mockRange);
			mockDocument.getText.mockReturnValue('unknown_fun');

			const canHandle = handler.canHandle(mockDocument, mockPosition);

			expect(canHandle).toBe(false);
		});

		test('returns true for known kfun', () => {
			mockDocument.getText.mockReturnValue('test_fun');

			const canHandle = handler.canHandle(mockDocument, mockPosition);

			expect(canHandle).toBe(true);
		});
	});

	describe('hover content generation', () => {
		test('generates no hover when no range', () => {
			mockDocument.getWordRangeAtPosition.mockReturnValue(null);
			mockDocument.getText.mockReturnValue('test_fun');

			const hover = handler.createHover(mockDocument, mockPosition);

			expect(hover).toBeNull();
		});

		test('generates no hover when no content', () => {
			mockDocument.getText.mockReturnValue('');

			const hover = handler.createHover(mockDocument, mockPosition);

			expect(hover).toBeNull();
		});

		test('generates hover for function with all properties', () => {
			mockDocument.getText.mockReturnValue('test_fun');

			const hover = handler.createHover(mockDocument, mockPosition);

			expect(hover).toMatchSnapshot();
		});

		test('generates hover for function without parameters', () => {
			mockDocument.getText.mockReturnValue('no_params_fun');

			const hover = handler.createHover(mockDocument, mockPosition);

			expect(hover).toMatchSnapshot();
		});
	});
});
