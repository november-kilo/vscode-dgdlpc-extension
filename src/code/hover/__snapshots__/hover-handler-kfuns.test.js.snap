// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KfunsHoverHandler hover content generation generates hover for function with all properties 1`] = `
{
  "contents": [
    mockConstructor {
      "appendCodeblock": [MockFunction],
      "appendMarkdown": [MockFunction],
      "isTrusted": true,
      "supportHtml": true,
      "supportThemeIcons": true,
      "value": "### test_fun

[test_fun](command:lpc.showKfunDoc?%5B%22test_fun%22%5D)

Test function description

**Parameters:**

- \`param1\`: First parameter
- \`param2\` (optional): Second parameter


**See Also:**[related_fun](command:lpc.showKfunDoc?%5B%22related_fun%22%5D), [another_fun](command:lpc.showKfunDoc?%5B%22another_fun%22%5D)",
    },
  ],
  "range": {
    "end": {
      "character": 8,
      "line": 0,
    },
    "start": {
      "character": 0,
      "line": 0,
    },
  },
}
`;

exports[`KfunsHoverHandler hover content generation generates hover for function without parameters 1`] = `
{
  "contents": [
    mockConstructor {
      "appendCodeblock": [MockFunction],
      "appendMarkdown": [MockFunction],
      "isTrusted": true,
      "supportHtml": true,
      "supportThemeIcons": true,
      "value": "### no_params_fun

[no_params_fun](command:lpc.showKfunDoc?%5B%22no_params_fun%22%5D)

Function without parameters

",
    },
  ],
  "range": {
    "end": {
      "character": 8,
      "line": 0,
    },
    "start": {
      "character": 0,
      "line": 0,
    },
  },
}
`;
